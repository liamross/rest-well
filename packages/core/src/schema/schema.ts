import type {z} from "zod";
import type {
  CombineObjects,
  CombineStrings,
  CombineZodSchemas,
  MediaType,
  Method,
  NoEmptyObject,
  Path,
  PathParams,
  PathParamSchema,
  Prettify,
  RequestBody,
  RequestHeaders,
  RequestQuery,
  Responses,
  ResponseWithHeaders,
  RestrictPath,
} from "../utils";
import type {Route, RouteProperties} from "./route";
import {combineObjects, combineStrings, combineZodSchemas} from "../utils";

/** Shared schema properties. */
type _SchemaShared<SR extends Responses | undefined = undefined, SH extends RequestHeaders | undefined = undefined> = {
  sharedResponses?: SR;
  sharedHeaders?: SH;
};

/** Properties passed into a schema builder. */
type SchemaCreateProperties<
  Res extends Schema,
  BP extends Path,
  BPP extends PathParamSchema<BP>,
  SR extends Responses | undefined,
  SH extends RequestHeaders | undefined,
> = {routes: Res} & (BPP extends undefined ? {} : {pathParams: NonNullable<BPP>}) & _SchemaShared<SR, SH>;

/** The properties defined for the schema. */
type SchemaProperties<
  BP extends Path,
  BPP extends PathParams | undefined,
  SR extends Responses | undefined,
  SH extends RequestHeaders | undefined,
> = {basePath: BP; pathParams: BPP} & _SchemaShared<SR, SH>;

/** Non-generic schema properties. For internal use. */
type _SchemaProperties = {
  basePath: string;
  pathParams?: PathParams;
  sharedResponses?: Responses;
  sharedHeaders?: RequestHeaders;
};

/** A non-generic basic schema object. This is usually for internal use only. */
export type Schema = {[key: string]: Route | Schema};

/** Result of combining schema properties with a route. */
type RoutePropertiesCombined<
  // Route.
  P extends Path,
  PP extends PathParams | undefined,
  M extends Method,
  R extends Responses,
  CT extends MediaType | undefined,
  B extends RequestBody | undefined,
  Q extends RequestQuery | undefined,
  H extends RequestHeaders | undefined,
  // Schema
  BP extends Path,
  BPP extends PathParams | undefined,
  SR extends Responses | undefined,
  SH extends RequestHeaders | undefined,
> = RouteProperties<
  CombineStrings<BP, P>,
  CombineZodSchemas<BPP, PP>,
  M,
  CombineObjects<SR, R>,
  CT,
  B,
  Q,
  CombineZodSchemas<SH, H>
>;

/** Combine schema properties with a route. */
function combineRouteWithSchema(route: Route, schemaProperties?: _SchemaProperties): Route {
  return {
    path: combineStrings(schemaProperties?.basePath, route.path),
    pathParams: combineZodSchemas(schemaProperties?.pathParams, route.pathParams),
    headers: combineZodSchemas(schemaProperties?.sharedHeaders, route.headers as RequestHeaders | undefined),
    responses: combineObjects(schemaProperties?.sharedResponses, route.responses),
    method: route.method,
    body: route.body,
    contentType: route.contentType,
    deprecated: route.deprecated,
    description: route.description,
    query: route.query,
    summary: route.summary,
  };
}

/** All routes combined with schema properties. */
export type FlushedSchema<
  Res extends Schema,
  BP extends Path,
  BPP extends PathParamSchema<BP>,
  SR extends Responses | undefined,
  SH extends RequestHeaders | undefined,
> = {
  [K in keyof Res]: Res[K] extends RouteProperties<
    infer P,
    infer PP,
    infer M,
    infer R,
    infer CT,
    infer B,
    infer Q,
    infer H
  >
    ? Prettify<RoutePropertiesCombined<P, PP, M, R, CT, B, Q, H, BP, BPP, SR, SH>>
    : Res[K] extends Schema
      ? FlushedSchema<Res[K], BP, BPP, SR, SH>
      : never;
};

/** Combine all routes with the properties passed to the schema. */
function flushSchema<
  Res extends Schema,
  BP extends Path,
  BPP extends PathParamSchema<BP>,
  SR extends Responses | undefined,
  SH extends RequestHeaders | undefined,
>(schema: Res, properties: SchemaProperties<BP, BPP, SR, SH>): FlushedSchema<Res, BP, BPP, SR, SH> {
  const flushed: Schema = {};
  for (const key in schema) {
    const value = schema[key]!;
    if (isRoute(value)) {
      flushed[key] = combineRouteWithSchema(value, properties);
    } else {
      flushed[key] = flushSchema(value, properties);
    }
  }
  return flushed as FlushedSchema<Res, BP, BPP, SR, SH>;
}

export function schema<
  Res extends Schema,
  SR extends Responses | undefined = undefined,
  SH extends RequestHeaders | undefined = undefined,
>(properties: SchemaCreateProperties<Res, "", undefined, SR, SH>): FlushedSchema<Res, "", undefined, SR, SH>;

export function schema<
  Res extends Schema,
  BP extends Path,
  BPP extends PathParamSchema<BP>,
  SR extends Responses | undefined = undefined,
  SH extends RequestHeaders | undefined = undefined,
>(
  basePath: RestrictPath<BP>,
  properties: SchemaCreateProperties<Res, BP, BPP, SR, SH>,
): FlushedSchema<Res, BP, BPP, SR, SH>;

export function schema<
  Res extends Schema,
  BP extends Path,
  BPP extends PathParamSchema<BP>,
  SR extends Responses | undefined = undefined,
  SH extends RequestHeaders | undefined = undefined,
>(
  pathOrProps: BP | SchemaCreateProperties<Res, BP, BPP, SR, SH>,
  propsIfPath?: SchemaCreateProperties<Res, BP, BPP, SR, SH>,
): FlushedSchema<Res, BP, BPP, SR, SH> | FlushedSchema<Res, "", undefined, SR, SH> {
  if (typeof pathOrProps === "string") {
    if (!propsIfPath) throw new Error("Second argument must be a schema properties object.");
    return flushSchema(propsIfPath.routes, getSchemaProps(pathOrProps, propsIfPath));
  }
  return flushSchema(pathOrProps.routes, getSchemaProps("", pathOrProps));
}

function getSchemaProps<
  Res extends Schema,
  BP extends Path,
  BPP extends PathParamSchema<BP>,
  SR extends Responses | undefined,
  SH extends RequestHeaders | undefined,
>(basePath: BP, props: SchemaCreateProperties<Res, BP, BPP, SR, SH>): SchemaProperties<BP, BPP, SR, SH> {
  const route = props as typeof props & {pathParams: BPP};
  return {...route, basePath};
}

export type RouteResponseValue<R extends Route> = {
  [k in keyof R["responses"]]: {status: k} & (R["responses"][k] extends z.ZodType<infer T>
    ? T extends undefined | void
      ? {}
      : {body: T}
    : R["responses"][k] extends ResponseWithHeaders
      ? {
          body: z.infer<R["responses"][k]["body"]>;
          headers: z.infer<R["responses"][k]["headers"]>;
        }
      : {});
}[keyof R["responses"]];

export type RouteRequestValue<R extends Route> =
  R extends RouteProperties<infer _P, infer PP, infer _M, infer _R, infer _CT, infer B, infer Q, infer H>
    ? TrimRoutePropertyKey<PP, "params"> &
        TrimRoutePropertyKey<B, "body"> &
        TrimRoutePropertyKey<Q, "query"> &
        TrimRoutePropertyKey<H, "headers">
    : {};

type TrimRoutePropertyKey<T, K extends string> = T extends z.ZodTypeAny
  ? NoEmptyObject<z.infer<T>> extends never
    ? {}
    : {[KK in K]: z.infer<T>}
  : {};

/**
 * Used to determine whether an object is a route.
 * @param route Either a route or a schema.
 * @returns True if the object is a route.
 */
export function isRoute<T>(route: T | Route): route is Route {
  return (
    typeof route === "object" &&
    route !== null &&
    "method" in route &&
    typeof route.method === "string" &&
    "path" in route &&
    typeof route.path === "string"
  );
}
