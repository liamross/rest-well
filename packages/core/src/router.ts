import type z from "zod";
import type {Resource, Route, RouteProperties, RouteReturnValue} from "./resource";
import type {InferZod, Prettify} from "./type-utils";

type RouteRequest<R extends Route> =
  R extends RouteProperties<infer _P, infer PP, infer _M, infer B, infer _CT, infer _R, infer Q, infer H>
    ? (PP extends z.ZodType<infer O> ? {params: Prettify<O>} : {}) &
        (InferZod<B> extends undefined ? {_body: InferZod<B>} : {body: Prettify<z.infer<B>>}) &
        (InferZod<Q> extends undefined ? {_query: InferZod<Q>} : {query: Prettify<z.infer<Q>>}) &
        (InferZod<H> extends undefined ? {_headers: InferZod<H>} : {headers: Prettify<z.infer<H>>})
    : never;

type RouteImplementation<R extends Resource | Route> = R extends Resource
  ? {[K in keyof R]: RouteImplementation<R[K]>}
  : R extends Route
    ? (req: Prettify<RouteRequest<R>>) => Promise<Prettify<RouteReturnValue<R>>>
    : never;

export function route<R extends Route>(route: R, implementation: RouteImplementation<R>): RouteImplementation<R> {
  return implementation;
}

export function router<R extends Resource>(
  resource: R,
  implementations: RouteImplementation<R>,
): RouteImplementation<R> {
  return implementations;
}
