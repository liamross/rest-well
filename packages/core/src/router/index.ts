export * from "./router";
export * from "./path";
