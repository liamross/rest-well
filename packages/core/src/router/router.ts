import type {Route, RouteImplementation, Schema} from "../schema";

export type Router = <T extends Schema | Route>(
  schema: T,
  implementations: RouteImplementation<T>,
) => RouteImplementation<T>;
