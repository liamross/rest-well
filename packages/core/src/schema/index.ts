export * from "./schema";
export * from "./route";
