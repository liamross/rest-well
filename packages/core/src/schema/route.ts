import type {
  MutationMethod,
  Path,
  PathParams,
  Prettify,
  RestrictPath,
  RouteBody,
  RouteContentType,
  RouteHeaders,
  RouteMethod,
  RoutePathParams,
  RouteQuery,
  RouteResponses,
} from "../utils";

type _RouteShared<R extends RouteResponses, Q extends RouteQuery | undefined, H extends RouteHeaders | undefined> = {
  responses: R;
  query?: Q;
  headers?: H;
  summary?: string;
  description?: string;
  deprecated?: true;
};

type RouteCreateProperties<
  P extends Path,
  PP extends PathParams<P>,
  M extends RouteMethod,
  R extends RouteResponses,
  CT extends RouteContentType | undefined,
  B extends RouteBody | undefined,
  Q extends RouteQuery | undefined,
  H extends RouteHeaders | undefined,
> = _RouteShared<R, Q, H> &
  (PP extends undefined ? {} : {pathParams: NonNullable<PP>}) &
  (M extends MutationMethod ? {body?: B; contentType?: CT} : {});

/**
 * These are the generic properties that are inside a route object.
 */
export type RouteProperties<
  P extends Path,
  PP extends RoutePathParams | undefined,
  M extends RouteMethod,
  R extends RouteResponses,
  CT extends RouteContentType | undefined,
  B extends RouteBody | undefined,
  Q extends RouteQuery | undefined,
  H extends RouteHeaders | undefined,
> = {method: M; path: P; pathParams: PP; body?: B; contentType?: CT} & _RouteShared<R, Q, H>;

function routeMethodFactory<M extends RouteMethod>(
  method: M,
): {
  <
    R extends RouteResponses,
    CT extends RouteContentType | undefined,
    B extends RouteBody | undefined,
    Q extends RouteQuery | undefined = undefined,
    H extends RouteHeaders | undefined = undefined,
  >(
    routeObj: RouteCreateProperties<"", undefined, M, R, CT, B, Q, H>,
  ): Prettify<RouteProperties<"", undefined, M, R, CT, B, Q, H>>;

  <
    P extends Path,
    PP extends PathParams<P>,
    R extends RouteResponses,
    CT extends RouteContentType | undefined,
    B extends RouteBody | undefined,
    Q extends RouteQuery | undefined = undefined,
    H extends RouteHeaders | undefined = undefined,
  >(
    path: RestrictPath<P>,
    properties: RouteCreateProperties<P, PP, M, R, CT, B, Q, H>,
  ): Prettify<RouteProperties<P, PP, M, R, CT, B, Q, H>>;
};

function routeMethodFactory<M extends RouteMethod>(method: M) {
  return <
    P extends Path,
    PP extends PathParams<P>,
    R extends RouteResponses,
    CT extends RouteContentType | undefined = undefined,
    B extends RouteBody | undefined = undefined,
    Q extends RouteQuery | undefined = undefined,
    H extends RouteHeaders | undefined = undefined,
  >(
    path: RestrictPath<P> | RouteCreateProperties<"", undefined, M, R, CT, B, Q, H>,
    properties?: RouteCreateProperties<P, PP, M, R, CT, B, Q, H>,
  ): RouteProperties<P, PP, M, R, CT, B, Q, H> | RouteProperties<"", undefined, M, R, CT, B, Q, H> => {
    if (typeof path === "string") {
      if (properties === undefined) throw new Error("Second argument must be route object.");
      const route = properties as typeof properties & {pathParams: PP};
      return {...route, path: path as P, method};
    }
    const route = path as typeof path & {pathParams: PP};
    return {...route, path: "" as P, method};
  };
}

export type Route = {method: RouteMethod; path: string; [key: string]: unknown};

/** Retrieve data from the server. Usually used to read a target entity. */
export const GET: ReturnType<typeof routeMethodFactory<"GET">> = routeMethodFactory("GET");

/** Send data to the server. Usually used to create a new entity. */
export const POST: ReturnType<typeof routeMethodFactory<"POST">> = routeMethodFactory("POST");

/** Replace data on the server. Usually used to create or overwrite a target. */
export const PUT: ReturnType<typeof routeMethodFactory<"PUT">> = routeMethodFactory("PUT");

/** Update data on the server. Usually used to update a target entity. */
export const PATCH: ReturnType<typeof routeMethodFactory<"PATCH">> = routeMethodFactory("PATCH");

/** Remove data from the server. Usually used to delete a target entity. */
export const DELETE: ReturnType<typeof routeMethodFactory<"DELETE">> = routeMethodFactory("DELETE");
