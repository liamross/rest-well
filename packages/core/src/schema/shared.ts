import type z from "zod";

// Path is just a string, but we enforce it using `RestrictPath`.
export type Path = string;

// Path params are more limited since they must be strings.
type PathParamValue = string | number | boolean | Date;

// We distinguish between query and mutation methods since mutation methods can
// have a body and content type.
export type QueryMethod = "GET";
export type MutationMethod = "POST" | "DELETE" | "PUT" | "PATCH";

export type ObjectSchema<T = unknown> = z.ZodType<{[key: string]: T}>;
export type AnySchema = z.ZodType<unknown>;

// These are all the base types for route properties.
export type RouteMethod = QueryMethod | MutationMethod;
export type RouteContentType = "application/json" | "multipart/form-data" | "application/x-www-form-urlencoded";
export type RoutePathParams = ObjectSchema<PathParamValue>;
export type RouteResponses = {[key: number]: AnySchema};
export type RouteBody = AnySchema;
export type RouteQuery = ObjectSchema;
export type RouteHeaders = ObjectSchema;

// Enforce that path params has a key for every variable in the path.
type _PathParamsInnerType = {[key: string]: PathParamValue};
type _PathParamsInner<S extends Path> = S extends `${infer _Start}{${infer Param}}${infer Rest}`
  ? {
      [k in
        | (Param extends `${string}/${string}` ? never : Param)
        // Allow any here since we don't care what the string is converted into by zod.
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        | keyof (_PathParamsInner<Rest> extends infer O extends _PathParamsInnerType ? O : {})]: any;
    }
  : undefined;

export type PathParams<S extends Path> =
  _PathParamsInner<S> extends infer O extends _PathParamsInnerType ? z.ZodType<O> : undefined;

// =============================================================================
// Enforce path string.
// -----------------------------------------------------------------------------

type InvalidPathError<Original extends string, Error extends string> = `${Original}|${Error}`;
type FormatErrorOrString<T extends string> = T extends InvalidPathError<infer _O, infer E> ? `Invalid path: ${E}` : T;

type StringMatches<S extends string, Match extends string, Error extends string> = S extends Match
  ? S
  : InvalidPathError<S, Error>;
type StringDoesNotMatch<S extends string, Match extends string, Error extends string> = S extends Match
  ? InvalidPathError<S, Error>
  : S;

type CombineErrors<Error extends InvalidPathError<string, string>, Maybe extends string> =
  Error extends InvalidPathError<infer O, infer EE>
    ? Maybe extends InvalidPathError<infer _O, infer ME>
      ? InvalidPathError<O, `${EE}, ${ME}`>
      : Error
    : Maybe;

type StartsWithSlash<T extends string> =
  T extends InvalidPathError<infer O, infer _E>
    ? CombineErrors<T, StringMatches<O, `/${string}`, "must start with '/'">>
    : StringMatches<T, `/${string}`, "must start with '/'">;

type DoesNotEndWithSlash<T extends string> =
  T extends InvalidPathError<infer O, infer _E>
    ? CombineErrors<T, StringDoesNotMatch<O, `${string}/`, "must not end with '/'">>
    : StringDoesNotMatch<T, `${string}/`, "must not end with '/'">;

export type RestrictPath<T extends string> = FormatErrorOrString<DoesNotEndWithSlash<StartsWithSlash<T>>>;
