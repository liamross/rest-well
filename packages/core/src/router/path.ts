import type {FlushedSchema, Route, RouteProperties, Schema} from "../schema";
import type {Prettify, PrettifyDeep, RouteMethod} from "../utils";
import type {RouterFunction, RouterImplementation} from "./router";
import {isRoute} from "../schema";

// Symbols to represent dynamic paths and methods.
const pathParam = Symbol("pathParam");
const GET = Symbol("GET");
const POST = Symbol("POST");
const DELETE = Symbol("DELETE");
const PUT = Symbol("PUT");
const PATCH = Symbol("PATCH");

// Translate a method string into its appropriate symbol.
const methodToSymbol = {
  /* eslint-disable @typescript-eslint/no-unnecessary-type-assertion */
  GET: GET as typeof GET,
  POST: POST as typeof POST,
  DELETE: DELETE as typeof DELETE,
  PUT: PUT as typeof PUT,
  PATCH: PATCH as typeof PATCH,
  /* eslint-enable @typescript-eslint/no-unnecessary-type-assertion */
} satisfies {[_ in RouteMethod]: symbol};

type SplitPath<P extends string> = P extends `${infer Head}/${infer Rest}`
  ? "" extends Head
    ? [...SplitPath<Rest>]
    : Head extends `{${string}}`
      ? [typeof pathParam, ...SplitPath<Rest>]
      : [Head, ...SplitPath<Rest>]
  : P extends `{${string}}`
    ? [typeof pathParam]
    : [P];

/** Turn a path string into an array of strings and pathParam symbols. */
function splitPath<P extends string>(path: P): SplitPath<P> {
  const segments = path.split("/").filter((s) => s !== "");
  return segments.map((s) => (s.startsWith("{") && s.endsWith("}") ? pathParam : s)) as SplitPath<P>;
}

type PathSegments = (string | typeof pathParam)[];

type SplitPathToObj<P extends PathSegments, Child> = P extends [
  infer K extends string | typeof pathParam,
  ...infer Rest extends PathSegments,
]
  ? {[_ in K]: Rest extends [] ? Child : SplitPathToObj<Rest, Child>}
  : Child;

// function splitPathToObj<P extends PathSegments, Child>(path: P, child: Child): SplitPathToObj<P, Child> {
//   if (path.length === 0) return child as SplitPathToObj<P, Child>;
//   const [head, ...rest] = path;
//   return {[head!]: splitPathToObj(rest, child)} as SplitPathToObj<P, Child>;
// }

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type UnionToIntersection<U> = (U extends any ? (x: U) => void : never) extends (x: infer I) => void ? I : never;

type RouteAndImplementation<R extends Route> = {
  [_ in (typeof methodToSymbol)[R["method"]]]: {route: R; fn: RouterFunction<R>};
};

// Flattened router input.

export type FlattenedRouterInput<T extends Schema | Route> = PrettifyDeep<
  T extends Schema
    ? UnionToIntersection<{[K in keyof T]: FlattenedRouterInput<T[K]>}[keyof T]>
    : T extends Route
      ? "" extends T["path"]
        ? RouteAndImplementation<T>
        : {[_ in T["path"]]: RouteAndImplementation<T>}
      : never
>;

type _RouterFunction = (...args: unknown[]) => unknown;
type _RouteAndImplementation = {route: Route; fn: _RouterFunction};
type _FlattenedRouterLeaf = {
  [GET]?: _RouteAndImplementation;
  [POST]?: _RouteAndImplementation;
  [DELETE]?: _RouteAndImplementation;
  [PUT]?: _RouteAndImplementation;
  [PATCH]?: _RouteAndImplementation;
};
type _FlattenedRouterInput = _FlattenedRouterLeaf & {[path: string]: _FlattenedRouterLeaf};

// Path tree.

export type PathTree<R extends Schema | Route> =
  R extends FlushedSchema<infer Res, infer BP, infer _BPP, infer _SR, infer _SH>
    ? SplitPathToObj<SplitPath<BP>, Prettify<UnionToIntersection<PathTree<Res[keyof Res]>>>>
    : R extends RouteProperties<infer P, infer _PP, infer M, infer _R, infer _CT, infer _B, infer _Q, infer _H>
      ? "" extends P
        ? {[_ in (typeof methodToSymbol)[M]]: RouterFunction<R & Route>}
        : SplitPathToObj<SplitPath<P>, {[_ in (typeof methodToSymbol)[M]]: RouterFunction<R & Route>}>
      : never;

type _PathTree = {
  [x: string]: _PathTree;
  [pathParam]?: _PathTree;

  [GET]?: _RouteAndImplementation;
  [POST]?: _RouteAndImplementation;
  [DELETE]?: _RouteAndImplementation;
  [PUT]?: _RouteAndImplementation;
  [PATCH]?: _RouteAndImplementation;
};

export function flattenRouterTree<S extends Schema>(
  schema: S,
  implementation: RouterImplementation<S>,
): FlattenedRouterInput<S> {
  return flattenRouterTreeInner(schema, implementation) as FlattenedRouterInput<S>;
}

function flattenRouterTreeInner(
  schema: Schema | Route,
  implementation: RouterImplementation<Schema | Route>,
): _FlattenedRouterInput {
  let result = {} as _FlattenedRouterInput;

  if (isRoute(schema)) {
    if (schema.path === "") {
      const method = methodToSymbol[schema.method];
      result[method] = {route: schema, fn: implementation as _RouterFunction};
    } else {
      if (!result[schema.path]) result[schema.path] = {};
      if (result[schema.path]![methodToSymbol[schema.method]]) {
        throw new Error(`Fatal: Duplicate route: ${schema.path} ${schema.method}`);
      }
      result[schema.path]![methodToSymbol[schema.method]] = {
        route: schema,
        fn: implementation as _RouterFunction,
      };
    }
  } else {
    for (const key in schema) {
      if (Object.prototype.hasOwnProperty.call(schema, key)) {
        const schemaOrRoute = schema[key]!;
        const childImplementation = implementation[key as keyof typeof implementation];
        if (!childImplementation) throw new Error(`Fatal: No implementation found for route: ${key}`);
        result = {...result, ...flattenRouterTreeInner(schemaOrRoute, childImplementation)};
      }
    }
  }

  return result;
}

function getChildTreeFromPath(pathTree: _PathTree, path: string): _PathTree {
  if (path !== "") return pathTree;
  const pathSegments = splitPath(path);

  let current = pathTree;
  for (const segment of pathSegments) {
    if (!current[segment]) current[segment] = {};
    current = current[segment]!;
  }

  return current;
}

function setMethodImplementations(oldTree: _PathTree, newTree: _FlattenedRouterLeaf): void {
  const keys = Object.getOwnPropertySymbols(newTree) as (typeof methodToSymbol)[keyof typeof methodToSymbol][];

  for (const key of keys) {
    const value = newTree[key]!;
    if (oldTree[key]) throw new Error(`Fatal: Duplicate route: ${value.route.path} ${value.route.method}`);
    oldTree[key] = value;
  }
}

export function buildPathTree<S extends Schema>(input: FlattenedRouterInput<S>): PathTree<S> {
  return buildPathTreeInner(input) as PathTree<S>;
}

function buildPathTreeInner(input: _FlattenedRouterInput): _PathTree {
  const pathTree = {} as _PathTree;

  // This will implement all symbols aka method implementations.
  setMethodImplementations(pathTree, input);

  // All paths are flattened and contain methods.
  const paths = Object.keys(input);
  for (const path of paths) {
    const childTree = getChildTreeFromPath(pathTree, path);
    const implementations = input[path]!;
    setMethodImplementations(childTree, implementations);
  }

  return pathTree;
}

//

//

//

type PathWithParams = {path: string; params: {[key: string]: string}};

/**
 * Match the url to a template path and extract the parameters.
 *
 * @example
 * getPathWithParams("/users/1", ["/users/{id}"]); // {path: "/users/{id}", params: {id: "1"}}
 */
export function getPathWithParams(pathname: string, paths: string[]): PathWithParams | null {
  let match: (PathWithParams & {pathParts: string[]}) | null = null;

  const urlParts = pathname.split("/");

  pathLoop: for (const path of paths) {
    const pathParts = path.split("/");

    // Ensure pathParts is same length as urlParts so we can do non-null assertions.
    if (pathParts.length !== urlParts.length) continue;

    const params: {[key: string]: string} = {};

    for (let i = 0; i < pathParts.length; i++) {
      const pathPart = pathParts[i]!;
      const urlPart = urlParts[i]!;
      const matchPart = match && match.pathParts[i]!;

      if (pathPart === urlPart) {
        // If current match is less specific, we can clear it.
        if (matchPart && isVariable(matchPart)) match = null;
        continue;
      } else if (isVariable(pathPart)) {
        // If current match is more specific, we can skip this path.
        if (matchPart && !isVariable(matchPart)) continue pathLoop;
        params[pathPart.slice(1, -1)] = urlPart;
        continue;
      }

      // If no match but also not a variable, we can skip this path.
      continue pathLoop;
    }

    // If we still have a match by this point, then there are two routes with
    // the same level of specificity and that is an error.
    if (match) throw new Error(`You have two routes that match the same URL: '${match.path}' and '${path}'`);
    match = {path, params, pathParts};
  }

  return match && {path: match.path, params: match.params};
}

function isVariable(pathPart: string) {
  return pathPart.startsWith("{") && pathPart.endsWith("}");
}
