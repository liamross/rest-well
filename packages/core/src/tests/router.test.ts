/* eslint-disable @typescript-eslint/require-await */
import {describe, expect, it} from "vitest";
import {z} from "zod";
import {createRouter} from "../";
import {api} from "./data/router";
import {apiResource} from "./data/schema";

expect.addSnapshotSerializer({
  serialize(val, config, indentation, depth, refs, printer) {
    if (val instanceof z.ZodType) {
      const typeName = (val._def as {[key: string]: string}).typeName;
      return `[${typeName}]`;
    }
    return printer(val, config, indentation, depth, refs);
  },
  test(val): boolean {
    return val instanceof z.ZodType;
  },
});

describe("router", () => {
  it("should match snapshot", () => {
    expect(createRouter(apiResource, api)).toMatchSnapshot();
  });
});
