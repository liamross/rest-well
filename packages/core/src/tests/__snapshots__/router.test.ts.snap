// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`router > should match snapshot 1`] = `
{
  "api": {
    Symbol(VARIABLE): {
      "healthcheck": {
        Symbol(GET): {
          "fn": [Function],
          "route": {
            "body": undefined,
            "contentType": undefined,
            "deprecated": undefined,
            "description": undefined,
            "headers": [ZodObject],
            "method": "GET",
            "path": "/api/{version}/healthcheck",
            "pathParams": [ZodObject],
            "query": undefined,
            "responses": {
              "200": [ZodObject],
              "401": [ZodString],
              "500": [ZodString],
            },
            "summary": undefined,
          },
        },
      },
      "teams": {
        Symbol(POST): {
          "fn": [Function],
          "route": {
            "body": [ZodObject],
            "contentType": undefined,
            "deprecated": undefined,
            "description": undefined,
            "headers": [ZodObject],
            "method": "POST",
            "path": "/api/{version}/teams",
            "pathParams": [ZodObject],
            "query": undefined,
            "responses": {
              "201": [ZodObject],
              "401": [ZodString],
              "500": [ZodString],
            },
            "summary": undefined,
          },
        },
        Symbol(VARIABLE): {
          Symbol(DELETE): {
            "fn": [Function],
            "route": {
              "body": undefined,
              "contentType": undefined,
              "deprecated": undefined,
              "description": undefined,
              "headers": [ZodObject],
              "method": "DELETE",
              "path": "/api/{version}/teams/{id}",
              "pathParams": [ZodObject],
              "query": undefined,
              "responses": {
                "200": [ZodUndefined],
                "401": [ZodString],
                "500": [ZodString],
              },
              "summary": undefined,
            },
          },
        },
      },
      "users": {
        Symbol(VARIABLE): {
          Symbol(GET): {
            "fn": [Function],
            "route": {
              "body": undefined,
              "contentType": undefined,
              "deprecated": undefined,
              "description": undefined,
              "headers": [ZodObject],
              "method": "GET",
              "path": "/api/{version}/users/{id}",
              "pathParams": [ZodObject],
              "query": undefined,
              "responses": {
                "200": [ZodObject],
                "401": [ZodString],
                "404": [ZodUndefined],
                "500": [ZodString],
              },
              "summary": undefined,
            },
          },
          Symbol(PATCH): {
            "fn": [Function],
            "route": {
              "body": [ZodObject],
              "contentType": undefined,
              "deprecated": undefined,
              "description": undefined,
              "headers": [ZodObject],
              "method": "PATCH",
              "path": "/api/{version}/users/{id}",
              "pathParams": [ZodObject],
              "query": undefined,
              "responses": {
                "201": [ZodObject],
                "401": [ZodString],
                "404": [ZodObject],
                "500": [ZodString],
              },
              "summary": undefined,
            },
          },
          Symbol(DELETE): {
            "fn": [Function],
            "route": {
              "body": undefined,
              "contentType": undefined,
              "deprecated": undefined,
              "description": undefined,
              "headers": [ZodObject],
              "method": "DELETE",
              "path": "/api/{version}/users/{id}",
              "pathParams": [ZodObject],
              "query": undefined,
              "responses": {
                "200": [ZodVoid],
                "401": [ZodString],
                "404": [ZodObject],
                "500": [ZodString],
              },
              "summary": undefined,
            },
          },
        },
        Symbol(GET): {
          "fn": [Function],
          "route": {
            "body": undefined,
            "contentType": undefined,
            "deprecated": undefined,
            "description": undefined,
            "headers": [ZodObject],
            "method": "GET",
            "path": "/api/{version}/users",
            "pathParams": [ZodObject],
            "query": [ZodObject],
            "responses": {
              "200": {
                "body": [ZodArray],
                "headers": [ZodObject],
              },
              "401": [ZodString],
              "500": [ZodString],
            },
            "summary": undefined,
          },
        },
        Symbol(POST): {
          "fn": [Function],
          "route": {
            "body": [ZodObject],
            "contentType": undefined,
            "deprecated": undefined,
            "description": undefined,
            "headers": [ZodObject],
            "method": "POST",
            "path": "/api/{version}/users",
            "pathParams": [ZodObject],
            "query": [ZodObject],
            "responses": {
              "201": [ZodObject],
              "401": [ZodString],
              "500": [ZodString],
            },
            "summary": undefined,
          },
        },
        Symbol(DELETE): {
          "fn": [Function],
          "route": {
            "body": undefined,
            "contentType": undefined,
            "deprecated": undefined,
            "description": undefined,
            "headers": [ZodObject],
            "method": "DELETE",
            "path": "/api/{version}/users",
            "pathParams": [ZodObject],
            "query": undefined,
            "responses": {
              "200": [ZodUndefined],
              "401": [ZodString],
              "500": [ZodString],
            },
            "summary": undefined,
          },
        },
      },
    },
  },
}
`;
